fix: internal image drag & drop changes src

nadirayasmeen opened the pull request:
Related Ticket: https://github.com/hannonhill/cascade/issues/4200

Description of Changes:
* Placeholder text

Pre-checks:
* [ ] Changelog entry added
* [ ] Tests have been added (if applicable)
* [ ] Branch prefixed with `feature/` for new features (if applicable)
* [ ] License headers added on new files (if applicable)

Review:
* [ ] Milestone set
* [ ] Review comments resolved

GitHub issues (if applicable):

Commit:
Fix to remove cms instance from src when repositioning internal images hannonhill/Cascade#4200

Instance domain is getting prepended when drag & dropping an internal image in the CMS

nadirayasmeen commented:
Basically, the instance domain is getting added when an image gets drag & dropped in the WYSIWYG (as mentioned in the issue). I think it's more of a browser behavior than something TinyMCE does. First tried a couple of things on the TinyMCE/WYSIWYG initialization options side none which really worked, including converting the URL with `convert_url`, `relative_url`, `remove_script_host` & a function to override `urlconverter_callback` - which works as far as updating the image source via the source code, but the image dialog pop up would still show the repositioned image as an external image with the instance domain prepended to it.

Updated the image plugin to directly process the src logic & determine if it's a link coming from the instance url, then that instance domain should get stripped out. It reuses the same logic as the `urlUtil.getLocationOrigin` on the Cascade side.

tomusiaka commented on src/plugins/image/src/main/js/core/Utils.js:
```diff
@@ -178,6 +179,19 @@ define(
+     */
+    var removeOriginFromSrc = function (src) {
+      if (!window.location.origin) {
+        window.location.origin = window.location.protocol + "//" + window.location.hostname + (window.location.port ? ':' + window.location.port : '');
```
Is this a fix for a specific browser? Updating window.location normally results in a redirect and a full screen re-render. Also, [the docu](https://developer.mozilla.org/en-US/docs/Web/API/Location/origin) says that it's a read-only property. Or is it a special `window` object since I see it's passed as a parameter?

nadirayasmeen replied:
I've only tried it in Chrome before, but the same drag & drop issue happens in Firefox. This method doesn't seem to be causing a redirect/full screen render. The only other instance I see the original method used is specifying the  Cascade spellchecker as an external spellchecker in the WYSIWYG options.

tomusiaka replied:
Ah I see. And because we are in a completely separate project, we can't use urlUtil.js.

tomusiaka commented on src/plugins/image/src/main/js/ui/Dialog.js:
```diff
@@ -62,6 +62,9 @@ define(
         src = imgElm ? dom.getAttrib(imgElm, 'src') : '';
 
+        // Workaround to remove instance origin if it's been added, i.e. image has been repositioned
+        src = Utils.removeOriginFromSrc(src);
```
Since the change is in `Dialog.js` does this mean that the user will need to open the insert/edit image dialog to fix the path after repositioning an image? What happens when user does not open the dialog? Should we execute this type of cleanup when user is submitting the content?

tomusiaka commented on src/plugins/image/src/main/js/core/Utils.js:
```diff
@@ -178,6 +179,19 @@ define(
+
+      var origin = window.location.origin;
+
+      return src.replace(origin, "");
```
Perhaps removing the origin from the beginning of the string would be more reliable in case it's an external URL that for some reason has this origin as a request parameter, for example if origin is `https://example.com` and the src is `https://another.com/get/image?name=something&for=https://example.com`.

nadirayasmeen replied:
Good point 👍

Commit:
Fix incorrect internal image src's hannonhill/Cascade#4200

nadirayasmeen commented:
Posting [changes so far](https://github.com/hannonhill/tinymce/pull/6/commits/7f1ecc202d1e47d0a8824bb5c85165f4edc2534f), because I'm having trouble figuring out an issue and @rgriffith, wondering if you had any suggestions: Through logging/debugging, it appears the "bad" src gets updated whenever the image gets repositioned (or more like whenever editor content is updated/reparsed like you mentioned). In the source code editor, the img src is corrected. 

However, if you load the image dialog at this point, then the [instance url is still in the src here](https://github.com/hannonhill/tinymce/blob/cascade/src/plugins/image/src/main/js/ui/Dialog.js#L63) for some reason; so the image dialog populates the external url tab. Not sure if it's timing issue(?) or why the popup is hanging onto the old src 😕

Commit:
Clean up remove instance method logic hannonhill/Cascade#4200

rgriffith commented on src/plugins/image/src/main/js/core/Nodes.js:
```diff
@@ -0,0 +1,64 @@
+     * Helper method that removes the instance URL from a given source, in case it was added by the browser, i.e. when an internal image is repositioned
+     */
+      var removeInstanceHostFromSrc = function (src) {
+        if (!window.location.origin) {
+          window.location.origin = window.location.protocol + "//" + window.location.hostname + (window.location.port ? ':' + window.location.port : '');
+        }
+
+        var origin = window.location.origin;
```
Can be replaced with a call to `getInstanceHost()`

nadirayasmeen replied:
Thanks! There's still a bit of code cleanup left to do.

rgriffith reviewed (changes requested):
Minor suggestion while I look into the issue regarding the dialog having "old data"

rgriffith commented:
> However, if you load the image dialog at this point, then the instance url is still in the src here for some reason; so the image dialog populates the external url tab. Not sure if it's timing issue(?) or why the popup is hanging onto the old src

I wonder if we also need to add something like:

```
editor.parser.addNodeFilter('img', Nodes.fixBadInternalSrc(editor));
```

To ensure the `src` is cleaned up when parsing AND serializing.

nadirayasmeen commented:
Thanks Ryan! I've added 

```
editor.parser.addNodeFilter('img', Nodes.fixBadInternalSrc(editor));
```

and saw the same behavior with that dialog. Looking at their docs, I'm trying to figure out the difference between the serializer & the parser node filter, because the sound like they essentially do the same thing (adding the filter to the parser). Weird 😄 :

Serializer:
> Adds a node filter function to the parser used by the serializer, the parser will collect the specified nodes by name and then execute the callback once it has finished parsing the document.

DomParser:

> Adds a node filter function to the parser, the parser will collect the specified nodes by name and then execute the callback once it has finished parsing the document.

I've tried adding a `nodeChanged()` call unsuccessfully, but wondering if there needs to be some other type of way to "announce" to observers that the src has been change programmatically?

Commit:
Fixing issues where img src wasn't getting updated opening the image popup  hannonhill/Cascade#4200

nadirayasmeen commented on src/plugins/image/src/main/js/ui/Dialog.js:
```diff
@@ -61,12 +61,16 @@ define(
         }
 
+        // Workaround to ensure any bad internal src is updated
+        dom.setAttrib(imgElm, 'src', src);
```
For some reason, setting the src again here works. I used this same line in the filter callback, as well as trying to replace that node and other experiments, none I was able to get to work yet 👎

nadirayasmeen commented:
@rgriffith, it's not pretty but the only way I've been able to get this working so far is via the "double-check" method we were discussing in Slack (updating via the filter as well as before the dialog loads) + [this line](https://github.com/hannonhill/tinymce/pull/6/commits/be0210957b4552e6d50bd51e55f05cce5189c929#r498545515).

Commit:
Adds node filter to the parser in addition to the serializer hannonhill/Cascade#4200

Commit:
Clarifying method names, etc. addresses CR  hannonhill/Cascade#4200

Commit:
Fixing QA bugs with external images hannonhill/Cascade#4200

- Fixes issue where external images src was blank when loading the image popup
- Fixes issue where initially inserting external image, the src is undefined
